jerry-core/ecma/base/ecma-helpers-errol.c

```diff
@@ -0,0 +1,240 @@
+ *   copies of the Software, and to permit persons to whom the Software is
+ *   furnished to do so, subject to the following conditions:
+ *
+ *   The above copyright notice and this permission notice shall be included in all
+ *   copies or substantial portions of the Software.
+ *
+ *   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *   SOFTWARE.
+ */
+
+#include <math.h>
+
+#include "config.h"
+#include "ecma-helpers.h"
+
+/** \addtogroup ecma ECMA
+ * @{
+ *
+ * \addtogroup ecmahelpers Helpers for operations with ECMA data types
+ * @{
+ */
+
+#if CONFIG_ECMA_NUMBER_TYPE == CONFIG_ECMA_NUMBER_FLOAT64
+
+/**
+ * Printing Floating-Point Numbers
+ *
+ * available at http://cseweb.ucsd.edu/~mandrysc/pub/dtoa.pdf
+ */
+
+/**
+ * Floating point format definitions
+ */
+#define FP_MAX DOUBLE_MAX
+#define FP_MIN DOUBLE_MIN
+#define fpnext(val) (nextafter ((val), INFINITY))
+#define fpprev(val) (nextafter ((val), -INFINITY))
+
+#define ERROL0_EPSILON 0.0000001
+#define ERROL1_EPSILON 8.77e-15
+
+/**
+ * High-precision data structure.
+ */
+struct hp_t
+{
+  ecma_number_t val; /**< value */
+  ecma_number_t off; /**< offset */
+};
+
+/**
+ * Normalize the number by factoring in the error.
+ */
+static inline void __attr_always_inline___
+hp_normalize (struct hp_t *hp) /**< [in, out] float pair */
+{
+  ecma_number_t val = hp->val;
+
+  hp->val += hp->off;
+  hp->off += val - hp->val;
+} /* hp_normalize */
+
+/**
+ * Multiply the high-precision number by ten.
+ */
+static inline void __attr_always_inline___
+hp_mul10 (struct hp_t *hp) /**< [in, out] high-precision number */
+{
+  ecma_number_t off, val = hp->val;
+
+  hp->val *= 10.0;
+  hp->off *= 10.0;
+
+  off = hp->val;
+  off -= val * 8.0;
+  off -= val * 2.0;
+
+  hp->off -= off;
+
+  hp_normalize (hp);
+} /* hp_mul10 */
+
+/**
+ * Divide the high-precision number by ten.
+ */
+static void
+hp_div10 (struct hp_t *hp) /**< [in, out] high-precision number */
+{
+  ecma_number_t val = hp->val;
+
+  hp->val /= 10.0;
+  hp->off /= 10.0;
+
+  val -= hp->val * 8.0;
+  val -= hp->val * 2.0;
+
+  hp->off += val / 10.0;
+
+  hp_normalize (hp);
+} /* hp_div10 */
+
+/**
+ * Errol0 double to ASCII conversion, guaranteed correct but possibly not optimal.
+ *
+ * @return the exponent
+ */
+inline int32_t __attr_always_inline___
+ecma_errol0_dtoa (ecma_number_t val, /**< ecma number */
+                  uint64_t *digits_p, /**< [out] digits */
+                  int32_t *num_of_digits_p) /**< [out] number of digits */
```

Comment:
A bit nitpicky, but I'd rather have this function return its digits and have the exponent as an out parameter. But that's just a "looks nicer to me" issue.